fix: dedupliate cataloging binary artifacts

wagoodman opened the pull request:
This PR makes the following adjustments:
- primarily restores the capability to reference binary artifacts when `sbom[].artifact == binary`
- adds the ability to have templates within `sbom[].env` values
- only prepends `dist` config value to paths that relative paths (takes no action on absolute paths)
- decomposes templating code further for the SBOM pipe and surround with more testing

Why make this change? primarily to unlock more workflows when generating SBOMs.

Not all tooling can work with encapsulations of go binaries (zip/tar) and instead need access to the binary directly. Since goreleaser manages creating the binary it is possible to expose the original binary to SBOM tooling even though it may be an encapsulation that is attached to a release (e.g. a zip or tar.gz)

Take for example using `cyclonedx-gomod app`:

```
#.goreleaser.yaml

archives:
  - format: tar.gz

sboms:
  - documents:
      - "{{ .Binary }}_{{ .Version }}_{{ .Os }}_{{ .Arch }}.cdx.sbom"
    artifacts: binary
    cmd: cyclonedx-gomod
    env: 
      - GOOS={{ .Os }}
      - GOARCH={{ .Arch }}
    args: ["app",  "-json", "-output", "$document", "$PWD"]

```

With `artifacts: binary` the SBOM tool gets access to the unarchived binary even though `archives[0].format = tar.gz` and there is **_no_** `archives[].format = binary` set. Additionally this uses goreleaser-sourced variables to seed the correct go-specific environment variables that the tool keys in on.

Relevant links:
- Support additional SBOM files based on CycloneDX: https://github.com/goreleaser/goreleaser/issues/2808
- Removal of binary from the SBOM pipe artifact filter: https://github.com/goreleaser/goreleaser/pull/2798

caarlos0 reviewed (approved):
LGTM, seems like its missing a go mod tidy though

caarlos0 commented:
Thanks for the PR, its looking good to me overall! 

The only pointers I have are:

- the deduplication thing maybe is something worth moving to `artifacts.go`
- there is a test failing
- it looks like its missing a `go mod tidy`

other than that, lgtm!

PS: sorry for the delay reviewing this, was focusing on other things.
Thanks!

wagoodman commented:
@caarlos0 no worries -- this one fell off my radar too.

I went ahead and moved the filter function to `artifacts.go`, fixed the failing tests, rebased + go mod tidied. Should be good to go 👍

wagoodman commented:
@caarlos0 nudge --let me know if there is anything else you'd like changed/checked on this PR.

caarlos0 commented on go.mod:
```diff
@@ -28,7 +28,8 @@ require (
 	github.com/muesli/roff v0.1.0
-	github.com/slack-go/slack v0.10.2
+	github.com/scylladb/go-set v1.0.2
+	github.com/slack-go/slack v0.10.1
```
```suggestion
	github.com/slack-go/slack v0.10.2
```

wagoodman replied:
curious! I've adjusted the commit to keep v0.10.2 of slack 👍

caarlos0 reviewed (approved):
lgtm except the accidental dep downgrade

caarlos0 commented:
Thanks!
